### ISSUE TITLE
Dev deps that conflict with subdeps can produce invalid trees

### ISSUE THREAD

**Issue by iarna** [npm/npm#11766]
Using this `package.json`:

```
{
  "dependencies": {
    "aws-sdk": "^2.2.41"
  },
  "devDependencies": {
    "lodash": "^4.6.1"
  }
}
```

```
$ npm install
example@1.0.0 /Users/rebecca/code/npmtest/example/
├─┬ aws-sdk@2.2.41
│ ├── sax@1.1.5
│ ├── xml2js@0.4.15
│ └── xmlbuilder@2.6.2
└── lodash@4.6.1
$ npm ls
example@1.0.0 /Users/rebecca/code/npmtest/example
├─┬ aws-sdk@2.2.41
│ ├── sax@1.1.5
│ ├── xml2js@0.4.15
│ └─┬ xmlbuilder@2.6.2
│   └── UNMET DEPENDENCY lodash@~3.5.0
└── lodash@4.6.1
```

Running `npm install` a second time corrects the issue.

**Comment by robey** [comment]
thanks for listening to me flail helplessly on this and finding a simple repro case. :)

![thanks](https://cloud.githubusercontent.com/assets/172/13514168/5847533e-e15b-11e5-88f3-813cf11c6d23.jpg)

**Comment by nmccready** [comment]
Yup this just bit us in the but here where we were using object-assign as a devDep and gulp needed it as a dependency. However Heroku nodejs buildpack strips (prunes) all devDependencies. So gulp would bomb out. 

Any idea of resolution? We are seeing this on 2.14.22

**Comment by munro** [comment]
I'm experiencing a very similar issue: `missing: async@^1.4.0, required by handlebars@4.0.5`

I believe what's happening is `devDependencies` is not being considered when flattening the packages, so `dependencies.handlebars` installs `1.x` to `node_modules/`, then `devDependencies.async` comes along and clobbers it with `2.x`.  So when I go to shrinkwrap, it sees this discrepancy and aborts.

Handlebars has their dependencies setup correctly, and when I poke into `node_modules/handlebars/node_modules/` I don't see an `async` package.

Immediate work around solution is moving the `devDependencies` to `dependencies`.

```json
{
  "dependencies": {
    "handlebars": "^4.0.5",
  },
  "devDependencies": {
    "async": "^2.0.0-rc.2",
  }
}
```

node v6.1.0
npm 3.8.6

**Comment by schmod** [comment]
This may be similar to the transitive dependency issue described in #10727.  

Does running `npm install` a second time produce a valid tree?

**Comment by robey** [comment]
just checking in. coworker got freshly bitten by this today, despite having our builds pinned to npm 2.

will happily try out potential fixes if you have some ideas and have trouble reproing.

**Comment by zkat** [comment]
This is quite probably fixed as of npm@4.0.5 thanks to #12811. Cheers!
